services/src/main/java/org/keycloak/broker/saml/mappers/XPathAttributeMapper.java

```diff
@@ -0,0 +1,305 @@
+    public static final String USER_ATTRIBUTE = "user.attribute";
+    private static final String EMAIL = "email";
+    private static final String FIRST_NAME = "firstName";
+    private static final String LAST_NAME = "lastName";
+    private static final Set<IdentityProviderSyncMode> IDENTITY_PROVIDER_SYNC_MODES = new HashSet<>(Arrays.asList(IdentityProviderSyncMode.values()));
+
+    static {
+        ProviderConfigProperty property;
+        property = new ProviderConfigProperty();
+        property.setName(ATTRIBUTE_XPATH);
+        property.setLabel("Attribute XPath");
+        property.setHelpText("XPath to search for in attribute.");
+        property.setType(ProviderConfigProperty.STRING_TYPE);
+        configProperties.add(property);
+        property = new ProviderConfigProperty();
+        property.setName(ATTRIBUTE_NAME);
+        property.setLabel("Attribute Name");
+        property.setHelpText("Name of attribute to search for in assertion and apply xpath. You can leave this blank to try to apply xpath to all attributes or specify a friendly name instead.");
+        property.setType(ProviderConfigProperty.STRING_TYPE);
+        configProperties.add(property);
+        property = new ProviderConfigProperty();
+        property.setName(ATTRIBUTE_FRIENDLY_NAME);
+        property.setLabel("Friendly Name");
+        property.setHelpText("Friendly name of attribute to search for in assertion. You can leave this blank to try to apply xpath to all attributes or specify a name instead.");
+        property.setType(ProviderConfigProperty.STRING_TYPE);
+        configProperties.add(property);
+        property = new ProviderConfigProperty();
+        property.setName(USER_ATTRIBUTE);
+        property.setLabel("User Attribute Name");
+        property.setHelpText("User attribute name to store xpath value.  Use email, lastName, and firstName to map to those predefined user properties.");
+        property.setType(ProviderConfigProperty.STRING_TYPE);
+        configProperties.add(property);
+    }
+
+    public static final String PROVIDER_ID = "saml-xpath-attribute-idp-mapper";
+
+    @Override
+    public boolean supportsSyncMode(IdentityProviderSyncMode syncMode) {
+        return IDENTITY_PROVIDER_SYNC_MODES.contains(syncMode);
+    }
+
+    @Override
+    public List<ProviderConfigProperty> getConfigProperties() {
+        return configProperties;
+    }
+
+    @Override
+    public String getId() {
+        return PROVIDER_ID;
+    }
+
+    @Override
+    public String[] getCompatibleProviders() {
+        return COMPATIBLE_PROVIDERS;
+    }
+
+    @Override
+    public String getDisplayCategory() {
+        return "Attribute Importer";
+    }
+
+    @Override
+    public String getDisplayType() {
+        return "XPath Attribute Importer";
+    }
+
+    @Override
+    public void preprocessFederatedIdentity(KeycloakSession session, RealmModel realm, IdentityProviderMapperModel mapperModel, BrokeredIdentityContext context) {
```

Comment:
Although this method gets passed session and realm, it doesn't use them, and the other objects are quite plain. This makes it a good candidate for a unit test.

Therefore, I'd like to see a unit test at least for this method and a happy path. This would allow me to explore the code a lot more simpler.

A starting point could be the following snippet of code. Please update accordingly, and add 1-2 cases to provide the code coverage you would like to see.

```
public class XPathAttributeMapperTest extends TestCase {

    @Test
    public void test() {
        IdentityProviderMapperModel mapperModel = new IdentityProviderMapperModel();
        Map<String, String> config = new HashMap<>();
        mapperModel.setConfig(config);
        config.put(XPathAttributeMapper.USER_ATTRIBUTE, "email");
        config.put(XPathAttributeMapper.ATTRIBUTE_NAME, "email");
        config.put(XPathAttributeMapper.ATTRIBUTE_XPATH, "email");
        BrokeredIdentityContext context = new BrokeredIdentityContext("xzy");
        AssertionType assertion = AssertionUtil.createAssertion("xzy", NameIDType.deserializeFromString("xzy"));
        AttributeStatementType statement = new AttributeStatementType();
        assertion.getAttributeStatements().add(statement);
        context.getContextData().put(SAMLEndpoint.SAML_ASSERTION, assertion);
        new XPathAttributeMapper().preprocessFederatedIdentity(null, null, mapperModel, context);
        assertNotNull(context.getEmail());
    }
}
```

Reply:
Thanks for adding this test. Please now mark this thread to be resolved.